### ISSUE TITLE
Key presses are not handled in some situations when `UI::InputBlocking::Block` is returned from `OnKeyPress`

### ISSUE THREAD

**Issue by XertroV** [openplanet-nl/issues#98]
Copy-pasting [from Discord](https://discord.com/channels/276076890714800129/424967293538402334/970073584536027156) (`#angelscript` channel).

*(Note: github seems to limit the vertical size of quotes; you might need to scroll the quotes to see the whole thing.)*

> [7:26 AM] XertroV: Hey -- I'm a first time plugin author and am looking for some help w/ an issue I'm running into with OnKeyPress. 
> Broadly, my goal is to publish a plugin that prevents 'Give up' from being triggered in certain game modes (like COTD).
> 
> I want to handle a key press (Delete) and block the input so it doesn't get handled by the game. This part works (well, most of the time; like 90%).
> (Initially I wanted to actually unbind and rebind the key, but it doesn't look like there's a good way to do this AFAICT from the docs.)
> 
> I'm running into a very specific bug. I'm finding that if I already have another key pressed beforehand (e.g., up arrow), then other key presses are not handled until up is released. Practically, this means:
> - car is driving and steering works
> - block a key press (VirtualKey::Delete)
> - car keeps moving forward but steering does not work (nor breaks)
> 
> It's actually a bit worse than that. If you have up pressed (or any combination of keys), then press Del, then press any other button, then release up (or that original combination): the controls stay locked up. This can be chained by pressing another button before releasing a pressed button. Everything goes back to normal only after all buttons are released or the TM window is defocused. The controls shown via the dashboard don't react to new key presses either; it just shows a static image of the controls as they were pressed when Del was pressed.
> 
> I had a look in Debugger > Profiler and I can't see any problems.
> 
> I'm logging key presses (print("Key (" + key + ") " + (down ? "pressed" : "released"));) and no log output is generated till all the keys are released (then it starts logging like expected again).
> 
> My code is here: https://github.com/XertroV/tm-never-give-up/blob/master/src/main.as
> I can make a recording too if that will help.
> 
> I'm happy to write up an issue for openplanet if it's a bug, but I wanted to check here first to see if I was missing something or there is a better way to handle key presses or implement my plugin.

> [7:54 AM] Banalian/Nailanab: i've run into the same trouble trying to make the exact same idea, my only lead was to find where the bindings are ingame :/
> [7:54 AM] Banalian/Nailanab: since this blocking solution isn't user friendly at all 🤷‍♂️

> [8:06 AM] Miss:
> > I'm running into a very specific bug. I'm finding that if I already have another key pressed beforehand (e.g., up arrow), then other key presses are not handled until up is released.
>
> Ah yes, I actually know about this, but I forgot to properly fix it. Please open a github issue for that if you can!
> [8:07 AM] Miss: It's a bug and I know what's causing it, I just need to figure out a good solution for it

**Comment by codecat** [comment]
Not gonna fix this anytime soon I'm afraid. This will, once again, for like the third time, require a complete rewrite of how input to the game is blocked. It's a big project.

**Comment by XertroV** [comment]
[rutenl on discord mentioned a bug](https://discord.com/channels/276076890714800129/424967293538402334/975909137357631508) that I'd guess is closely related to this issue but is not explicitly mentioned above:

> [9:54 AM] rutenl: it seems if you block a down event you dont get the up event
> [9:54 AM] rutenl: is that intended?

It's implicitly mentioned above via:

> I'm logging key presses (print("Key (" + key + ") " + (down ? "pressed" : "released"));) and no log output is generated till all the keys are released (then it starts logging like expected again).

**Comment by codecat** [comment]
This might be fixable separately from the primary issue mentioned in this thread, can you open a new issue for that?

**Comment by CatzHoekk** [comment]
Isn´t this the same issue? 

If you block an event you don´t get any new events (and effectively block everything) until you released every key, regardless when you started pressing it, before blocking or after blocking. The first input that goes through again is a keydown after everything else has been released at the same time.


Here´s a  video of the behaviour.
https://cdn.discordapp.com/attachments/424967293538402334/976851247854403634/Trackmania_2022.05.19_-_16.10.45.09.mp4
https://discord.com/channels/276076890714800129/424967293538402334/976851253105676388

**Comment by codecat** [comment]
The way I see it, there are 2 separate issues:

1. Blocking a key from `OnKeyPress` works 90% of the time, where sometimes the key still gets passed to the game. (Probably a race condition)
2. While blocking 1 specific key, other keys won't be sent to the plugin.

What you are describing sounds like the second issue, right?

**Comment by CatzHoekk** [comment]
Yes, personally i have not ran into the first case (yet?).

> 2\. While blocking 1 specific key, other keys won't be sent to the plugin.

Or the game.
